clients/vault/src/metrics.rs

```diff
@@ -1,27 +1,522 @@
+		},
+	};
+	buffer.clear();
+	res
+}
+
+pub async fn metrics_handler() -> Result<impl Reply, Rejection> {
+	let mut metrics = serialize(&REGISTRY.gather());
+	let custom_metrics = serialize(&gather());
+	metrics.push_str(&custom_metrics);
+	Ok(metrics)
+}
+
+fn raw_value_as_currency(value: u128, _currency: CurrencyId) -> Result<f64, ServiceError<Error>> {
+	let scaling_factor = ONE;
+	Ok(value as f64 / scaling_factor)
+}
+
+pub async fn publish_locked_collateral<P: VaultRegistryPallet>(
+	vault: &VaultData,
+	parachain_rpc: P,
+) -> Result<(), ServiceError<Error>> {
+	if let Ok(actual_collateral) =
+		parachain_rpc.get_vault_total_collateral(vault.vault_id.clone()).await
+	{
+		let actual_collateral =
+			raw_value_as_currency(actual_collateral, vault.vault_id.collateral_currency())?;
+		vault.metrics.locked_collateral.set(actual_collateral);
+	}
+	Ok(())
+}
+
+pub async fn publish_required_collateral<P: VaultRegistryPallet>(
+	vault: &VaultData,
+	parachain_rpc: P,
+) -> Result<(), ServiceError<Error>> {
+	if let Ok(required_collateral) =
+		parachain_rpc.get_required_collateral_for_vault(vault.vault_id.clone()).await
+	{
+		let required_collateral =
+			raw_value_as_currency(required_collateral, vault.vault_id.collateral_currency())?;
+		vault.metrics.required_collateral.set(required_collateral);
+	}
+	Ok(())
+}
+
+pub async fn publish_collateralization<P: VaultRegistryPallet>(
+	vault: &VaultData,
+	parachain_rpc: P,
+) {
+	// if the collateralization is infinite, return 0 rather than logging an error, so
+	// the metrics do change in case of a replacement
+	let collateralization = parachain_rpc
+		.get_collateralization_from_vault(vault.vault_id.clone(), false)
+		.await
+		.unwrap_or(0u128);
+	let float_collateralization_percentage = FixedU128::from_inner(collateralization).to_float();
+	vault.metrics.collateralization.set(float_collateralization_percentage);
+}
+
+async fn publish_stellar_balance<P: OraclePallet>(parachain_rpc: P, vault: &VaultData) {
+	match vault.stellar_wallet.read().await.get_balance().await {
+		Ok(balance) => {
+			let currency_id = vault.vault_id.wrapped_currency();
+			let asset: Result<stellar::Asset, _> = currency_id.try_into();
+			let mut wrapped_to_collateral: u128 = 0;
+			if let Ok(asset) = asset {
+				let asset_balance = get_balance_for_asset(asset, balance);
+				if let Some(b) = asset_balance {
+					wrapped_to_collateral = parachain_rpc
+						.wrapped_to_collateral(b as u128, currency_id)
+						.await
+						.unwrap_or_else(|e| {
+							// unexpected error, but not critical so just continue
+							tracing::warn!("Failed to get balance: {}", e);
+							0
+						});
+				};
+			} else {
+				tracing::warn!("Incorrect stellar asset type");
+			}
+			vault.metrics.asset_balance.actual.set(wrapped_to_collateral as f64);
+		},
+		Err(e) => {
+			// unexpected error, but not critical so just continue
+			tracing::warn!("Failed to get balance: {}", e);
+			vault.metrics.asset_balance.actual.set(0 as f64);
```

Comment:
Something has to be wrong with the current setup, because when I run it, I see that the Stellar balance shown in metrics is set to 0.

Reply:
i found the problem.
there is no rpc method` oracle_wrappedToCollateral `
we did not expose any RPC for oracle pallet API
do we need to expose it? what do you think?
@ebma

Reply:
Now the error reproduced on test `spacewalk-standalone`
but the reason seems that there is no price feed for currency id
`AssetTypeCreditAlphanum4(AlphaNum4 { asset_code: [85, 83, 68, 67], issuer: PublicKeyTypeEd25519([20, 209, 150, 49, 176, 55, 23, 217, 171, 154, 54, 110, 16, 50, 30, 226, 102, 231, 46, 199, 108, 171, 97, 144, 240, 161, 51, 109, 72, 34, 159, 139]) })`
and error is 
`Err(Rpc(ClientError(Call(Custom(ErrorObject { code: InternalError, message: "Execution error: Module(ModuleError { index: 19, error: [0, 0, 0, 0], message: None })", data: None })))))`

Reply:
But we do have the price for USD so the problem can't actually be that there is no price, or am I wrong? Is some conversion buggy?

Reply:
@ebma 
Parachain status is offline. So error means that there is no fresh price feed for 
USDC
It is required currency id to be online for all parachain 
AssetTypeCreditAlphanum4(AlphaNum4 { asset_code: [85, 83, 68, 67], issuer: PublicKeyTypeEd25519([20, 209, 150, 49, 176, 55, 23, 217, 171, 154, 54, 110, 16, 50, 30, 226, 102, 231, 46, 199, 108, 171, 97, 144, 240, 161, 51, 109, 72, 34, 159, 139]) })

Reply:
Ah yeah it's probably because there is an issue with the testchain. I'll create a new ticket. But for now, if you are testing locally you can quick-fix the testchain by changing the XCMCurrencyConversion in `testchain/runtime/src/lib.rs` to this

```
impl XCMCurrencyConversion for SpacewalkNativeCurrencyKey {
	fn convert_to_dia_currency_id(token_symbol: u8) -> Option<(Vec<u8>, Vec<u8>)> {
		if token_symbol == 0 {
			return Some((b"Kusama".to_vec(), b"KSM".to_vec()))
		}

		// We assume that the blockchain is always 0 and the symbol represents the token symbol
		let blockchain = vec![0u8];
		let symbol = vec![token_symbol];

		Some((blockchain, symbol))
	}

	fn convert_from_dia_currency_id(blockchain: Vec<u8>, symbol: Vec<u8>) -> Option<u8> {
		// We assume that the blockchain is always 0 and the symbol represents the token symbol
		if blockchain.len() != 1 && blockchain[0] != 0 || symbol.len() != 1 {
			return None
		}
		return Some(symbol[0])
	}
}
```

Reply:
@ebma 
I tested to bridge tokens with Solar Wallet
![FCC6AD0A-E155-4D22-8ECB-7D8DF65DFAB2](https://user-images.githubusercontent.com/102041955/228948409-300fe4eb-b57a-44f0-9c51-b5a136677fb1.jpeg)

and after issuing the token on standalone side it shows
```
stellar_balance{currency="XCM(0)_\"USDC\":\"GAKNDFRRWA3RPWNLTI3G4EBSD3RGNZZOY5WKWYMQ6CQTG3KIEKPYWAYC\"",type="actual"} 112000000
stellar_balance{currency="XCM(0)_\"USDC\":\"GAKNDFRRWA3RPWNLTI3G4EBSD3RGNZZOY5WKWYMQ6CQTG3KIEKPYWAYC\"",type="required_lowerbound"} 120000000
stellar_balance{currency="XCM(0)_\"USDC\":\"GAKNDFRRWA3RPWNLTI3G4EBSD3RGNZZOY5WKWYMQ6CQTG3KIEKPYWAYC\"",type="required_upperbound"} 120000000
```